pkg/blocks/step.go

```diff
@@ -22,241 +22,176 @@ package blocks
+		if err != nil {
+			return err
 		}
-	case "linux":
-		if runtime.GOOS == "linux" {
-			return true, nil
+		if useDefaultCleanup {
+			if dca := s.action.GetDefaultCleanupAction(); dca != nil {
+				s.cleanup = dca
+				return nil
+			} else {
+				return fmt.Errorf("`cleanup: default` was specified but step %v is not an action type that has a default cleanup action", s.Name)
+			}
 		}
-	// Run even if a previous step has failed.
-	case "always":
-		return true, nil
 
-	default:
-		return false, nil
+		s.cleanup, err = s.ParseAction(&csf.CleanupSpec)
+		if err != nil {
+			return err
+		}
 	}
-	return false, nil
+	return nil
 }
 
-// MakeCleanupStep creates a CleanupAct based on the given yaml.Node.
-// If the node is empty or invalid, it returns nil. If the node contains a
-// BasicStep or FileStep, the corresponding CleanupAct is created and returned.
-//
-// **Parameters:**
-//
-// node: A pointer to a yaml.Node containing the parameters to
-// create the CleanupAct.
-//
-// **Returns:**
-//
-// CleanupAct: The created CleanupAct, or nil if the node is empty or invalid.
-//
-// error: An error if the node contains invalid parameters.
-func (a *Act) MakeCleanupStep(node *yaml.Node) (CleanupAct, error) {
-	if node.IsZero() {
-		return nil, nil
-	}
-
-	basic, berr := a.tryDecodeBasicStep(node)
-	if berr == nil && !basic.IsNil() {
-		logging.L().Debugw("cleanup step found", "basicstep", basic)
-		return basic, nil
-	}
+func (s *Step) Execute(execCtx TTPExecutionContext) (*ActResult, error) {
+	return s.action.Execute(execCtx)
+}
 
-	file, ferr := a.tryDecodeFileStep(node)
-	if ferr == nil && !file.IsNil() {
-		logging.L().Debugw("cleanup step found", "filestep", file)
-		return file, nil
+func (s *Step) Cleanup(execCtx TTPExecutionContext) (*ActResult, error) {
+	if s.cleanup != nil {
+		return s.cleanup.Execute(execCtx)
 	}
-
-	err := fmt.Errorf("invalid parameters for cleanup steps with basic [%v], file [%v]", berr, ferr)
-	logging.L().Errorw(err.Error(), zap.Error(err))
-	return nil, err
+	logging.L().Infof("No Cleanup Action Defined for Step %v", s.Name)
+	return &ActResult{}, nil
 }
 
-func (a *Act) tryDecodeBasicStep(node *yaml.Node) (*BasicStep, error) {
-	basic := NewBasicStep()
-	err := node.Decode(&basic)
-	if err == nil && basic.Name == "" {
-		basic.Name = fmt.Sprintf("cleanup-%s", a.Name)
-		basic.Type = StepCleanup
+func (s *Step) Validate(execCtx TTPExecutionContext) error {
+	if err := s.action.Validate(execCtx); err != nil {
+		return err
+	}
+	if s.cleanup != nil {
+		if err := s.cleanup.Validate(execCtx); err != nil {
+			return err
+		}
 	}
-	return basic, err
+	return nil
 }
 
-func (a *Act) tryDecodeFileStep(node *yaml.Node) (*FileStep, error) {
-	file := NewFileStep()
-	err := node.Decode(&file)
-	if err == nil && file.Name == "" {
-		file.Name = fmt.Sprintf("cleanup-%s", a.Name)
-		file.Type = StepCleanup
+func (s *Step) ParseAction(node *yaml.Node) (Action, error) {
```

Comment:
decoding both step actions and cleanup actions now all uses the same code, drastically simplifying things